### ISSUE TITLE
Expecting 404, getting 500: "Trying to access array offset on value of type bool"

### ISSUE THREAD

**Issue by erfanimani** [magento/magento2#31984]
### Preconditions (*)

1. Magento 2.4.1
2. PHP 7.4

### Steps to reproduce (*)

1. As an admin, create an saleable product with any URL, such as "box"
2. Set "Generate "category/product" URL Rewrites" to "No"
3. As a customer, visit the URL /abc/box.html

### Expected result (*)

1. 404 page not found error, as the category abc does not exist

### Actual result (*)

1. 500, Trying to access array offset on value of type bool on vendor/magento/module-catalog-url-rewrite/Model/Storage/DynamicStorage.php on line 182

PHP 7.4 is become stricter than PHP 7.3 with regards to these types of errors and that's why this bug is showing up it's head now.

https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/CatalogUrlRewrite/Model/Storage/DynamicStorage.php#L182

The way the code is written doesn't make sense: the false check is done after an array access is attempted; the false check should be done first.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by erfanimani** [comment]
@magento give me 2.4-develop instance

**Comment by erfanimani** [comment]
I can't seem to reproduce this. Are the develop instances running PHP 7.4?

**Comment by bluemwhitew** [comment]
@magento give me 2.4-develop instance with env php 7.4

**Comment by erfanimani** [comment]
@magento give me 2.4-develop instance with env php 7.4

**Comment by erfanimani** [comment]
@magento give me 2.4-develop instance with env php 7.4

**Comment by erfanimani** [comment]
I can't reproduce through the deployment service. I'm unsure of whether it's actually giving me a php 7.4 instance.

I can reproduce on 2.4-develop and 2.4.2 contributor package on my local.

**Comment by engcom-Bravo** [comment]
Hello @erfanimani 

Thank you for your report.

We are able to reproduce the issue on the latest 2.4-develop. 
Our PHP version is 7.4.15

Visiting the _http://mage2.loc/abc/box.html_ page leads to :x: _500 (Internal Server Error)_ and exception in log file
_[2021-02-15 13:33:51] main.CRITICAL: Notice: Trying to access array offset on value of type bool in /.../app/code/Magento/CatalogUrlRewrite/Model/Storage/DynamicStorage.php on line 182 {"report_id":"f1230a5621627f2350158dc2a8e9ae41561bf3b1c818c89e4ecdb38d3087d1ce","exception":"[object] (Exception(code: 0): Notice: Trying to access array offset on value of type bool in /.../app/code/Magento/CatalogUrlRewrite/Model/Storage/DynamicStorage.php on line 182 at /.../lib/internal/Magento/Framework/App/ErrorHandler.php:61)"} []_

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-40897` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by iphigenie** [comment]
thanks for reporting this @erfanimani I've been having this but I assumed a third party module was responsible

**Comment by gabrieldagama** [comment]
Hi @erfanimani. Thank you for your report. 
The issue has been fixed in magento/magento2#31985 by @erfanimani in 2.4-develop branch
Related commit(s):
 - [6316620cca1482d4f56fd4f1e5d3482e6b1f6deb](https://github.com/magento/magento2/commit/6316620cca1482d4f56fd4f1e5d3482e6b1f6deb)
 - [6da61dffbfd21df167192c9ae4a528ffe682024b](https://github.com/magento/magento2/commit/6da61dffbfd21df167192c9ae4a528ffe682024b)
 - [ae62eef387aef2bc925db0e6cc81aeeb2e9f2b27](https://github.com/magento/magento2/commit/ae62eef387aef2bc925db0e6cc81aeeb2e9f2b27)

The fix will be available with the upcoming 2.4.3 release.

**Comment by alucardatem** [comment]
Also .. there is an extra fix that is needed on the findProductRewritesByFilter

in between 
`       $categoryFromDb = $this->connection->fetchRow($this->prepareSelect($data));

        foreach ($productsFromDb as $productFromDb) {`

add `  if(!is_array($categoryFromDb) ) {
                return $rewrites;
            }` otherwise the error still pops up

**Comment by hostep** [comment]
@alucardatem: your remark makes sense I think, are you interested in sending in a PR for this?

**Comment by ihor-sviziev** [comment]
Hi @alucardatem,
Could you create a separate issue for that, or even if you have a fix - create a pull request?

Thank you!

**Comment by ihor-sviziev** [comment]
@alucardatem, just checked, looks like this code is never gets called due to the following code, which is removes category id: 
https://github.com/magento/magento2/blob/c0ab1d9c7a3ee84b1c880ee3fb2d16793a7edb9d/app/code/Magento/Catalog/Model/Product/Url.php#L160-L161
